wordpress-mobile/WordPress-Android
# [Site Name] Add tracks events

Fixes #16291

Adds the track events for the Site Name Screen & Experiment.

<img alt="track_events_logs_preview" src="https://user-images.githubusercontent.com/4588074/163576400-12aa5f19-aa06-4884-9ec3-657b04fa10d6.png">

**Note:** A part of the changes of this pull request are from #16335, that got merged to this one:
- Unit tests
- Calling the track events at the right time
- Fixes to the `SiteCreationTracker` methods for tracking Site Name events

---

To test:

> **Please note:** If testing this PR via Android Studio, you can verify Tracks events via the `Logcat` tool window.
If you're testing this on a PR build or a beta build of the app, you can either use the Tracks dashboard or the in-app logs (tap profile pic → `Help and Support` → `Application log`).
> 
> In this case, the **verification** steps from below are to be read as:
> > **Verify** that the `{event_name_from_log}` event is emitted (with the expected properties).

<details>
<summary>1. Toggle the Site Name feature <b>on</b></summary>

1. From My Site Go to Me → App Settings → Debug settings
2. Scroll to Features in development section and tap on SiteNameFeatureConfig
3. Tap RESTART THE APP button
4. Since this feature is A/B tested go to **Abacus**, find the `wpandroid_site_name_v1` experiment and assign to your test a8c-user the treatment variation

**Alternatively hardcode the feature on by returning `true` [here](https://github.com/wordpress-mobile/WordPress-Android/blob/521e6b3a296339bec026f8ee2648eab326d31129/WordPress/src/main/java/org/wordpress/android/util/config/SiteNameFeatureConfig.kt#L19)** 
</details>

2. Start the Site Creation flow
3. Verify in the logs (via Logcat in Android Studio) that the following entry logged:
   ```
   🔵 Tracked: enhanced_site_creation_site_name_experiment, Properties: {"variation":"treatment"}
   ```
   **Reminder** The `variation` property should default to `control` if not using an a8c-user and instead overriding the return value [here](https://github.com/wordpress-mobile/WordPress-Android/blob/521e6b3a296339bec026f8ee2648eab326d31129/WordPress/src/main/java/org/wordpress/android/util/config/SiteNameFeatureConfig.kt#L19).
4. Choose an intent for your site
5. `Verify` that the `Site name` screen is shown and the following entry is logged:
   ```
   🔵 Tracked: enhanced_site_creation_site_name_viewed
   ```
6. Tap `SKIP` in the top right of the app
7. `Verify` that the following entry is logged:
   ```
   🔵 Tracked: enhanced_site_creation_site_name_skipped
   ```
8. Navigate back ←
9. `Verify` that the entry at step **5** is not logged again
10. Enter a site name (e.g. `A Cool Site Name`) & tap `Continue` or Enter/Done on the keyboard
12. `Verify` that the following entry is logged:
   ```
   🔵 Tracked: enhanced_site_creation_site_name_entered, Properties: {"site_name":"A Cool Site Name"}
   ```

Default / (Control variation)
1. Toggle the Site Name feature <b>off</b> (see step 1. above)
   - Alternatively you can logout & login with a non a8c WordPress account to get assigned the default `Control` variation of the Site Name Experiment.
2. Start the Site Creation flow
3. Verify in the logs (via Logcat in Android Studio) that the following entry logged:
   ```
   🔵 Tracked: enhanced_site_creation_site_name_experiment, Properties: {"variation":"control"}
   ```
4. Choose an intent for your site
5. `Verify` that the `Choose a design` screen is shown and there is no additional logging that starts with the `enhanced_site_creation_site_name_` prefix.



## Regression Notes
1. Potential unintended areas of impact
   Site name screen user interaction (though unlikely)

3. What I did to test those areas of impact (or what existing automated tests I relied on)
   Manual testing

5. What automated tests I added (or what prevented me from doing so)
   @mkevins Added unit tests for the tracks events in `SiteCreationSiteNameViewModelTest` & `SiteCreationMainVMTest` & I merged his collab PR #16335 🙇‍♂️.

PR submission checklist:
